Cubical/Codata/M/AsLimit/Coalg/Properties.agda

```diff
@@ -0,0 +1,207 @@
+    missing-0-Iso : Iso (Lift {ℓ-zero} {ℓ} Unit) (Σ (Cone₀ C,γ) (λ u → u ≡ ϕ₀ u))
+    fun (missing-0-Iso) = (λ _ → u0 , (funExt p0))
+    inv (missing-0-Iso) = (λ x → lift tt)
+    rightInv (missing-0-Iso) = (λ b → missing-0-helper b)
+    leftInv (missing-0-Iso) = λ a i → lift tt
+
+    postulate
+      missing-2-Iso :
+        (x : Lift Unit)
+        → Iso (Lift {ℓ-zero} {ℓ} Unit)
+               (Σ[ q ∈ (Cone₁ C,γ (fun missing-0-Iso x .fst)) ] (PathP (λ i → Cone₁ C,γ (fun missing-0-Iso x .snd i)) q (ϕ₁ (fun missing-0-Iso x .fst) q)))
+
+U-contr : ∀ {ℓ} {S : Container ℓ} (C,γ : Coalg₀ S) -> isContr (U C,γ)
+U-contr {ℓ} C,γ = inv (contr-is-ext-Iso {A = U C,γ} (U-is-Unit-Iso C,γ)) (lift tt , λ { (lift tt) -> refl })
+  where
+    isContrIsPropPath : ∀ {ℓ} {A : Set ℓ} → (x : isContr A) → ∀ y → isProp (x .fst ≡ y)
+    isContrIsPropPath {A = A} x y = isContr→isProp (isContr→isContrPath x (x .fst) y)
+
+    contr-is-ext-Iso-helper :
+      ∀ {ℓ} {A B : Set ℓ}
+      → (p : Iso A B)
+      → ((a : A) → Iso (∀ y → a ≡ y) (∀ y → (fun p a) ≡ y))
+    fun (contr-is-ext-Iso-helper (iso f g rightI leftI) a) x y = cong f (x (g y)) ∙ rightI y
+    inv (contr-is-ext-Iso-helper (iso f g rightI leftI) a) x y = sym (leftI a) ∙ cong g (x (f y)) ∙ leftI y
+    rightInv (contr-is-ext-Iso-helper p@(iso f g rightI leftI) a) b = funExt λ y → isContrIsPropPath (f a , b) y (cong f (sym (leftI a) ∙ cong g (b (f (g y))) ∙ leftI (g y)) ∙ rightI y) (b y)
+    leftInv (contr-is-ext-Iso-helper p@(iso f g rightI leftI) a) b = funExt λ y → isContrIsPropPath (a , b) y (sym (leftI a) ∙ cong g (cong f (b (g (f y))) ∙ rightI (f y)) ∙ leftI y) (b y)
+
+    -- Can this be generalized to Iso A B → Iso (H A) (H B) , not just for H = isContr ?
+    contr-is-ext-Iso : ∀ {ℓ} {A B : Set ℓ} -> Iso A B -> Iso (isContr A) (isContr B)
+    contr-is-ext-Iso {A = A} {B} p = Σ-ap-iso p (contr-is-ext-Iso-helper p)
+
+----------------------------------------------------
+-- Finality properties for bisimulation relations --
+----------------------------------------------------
+
+lim-terminal : ∀ {ℓ} {S : Container ℓ} {C,γ : Coalg₀ S} → isContr (C,γ ⇒ M-coalg)
+lim-terminal {C,γ = C,γ} = inv (U-is-Unit-Iso C,γ) (lift tt) , U-contr C,γ .snd
+
+coalg-unfold : ∀ {ℓ} {S : Container ℓ} -> (C,γ : Coalg₀ S) -> (_⇒_ {S = S} (C,γ) (M-coalg {S = S}))
+coalg-unfold C,γ = lim-terminal {C,γ = C,γ} .fst
+
+-- unique function into final coalg
+coalg-unfold-universal : ∀ {ℓ} {S : Container ℓ} -> (C,γ : Coalg₀ S) -> (y : C,γ ⇒ M-coalg) → fst lim-terminal ≡ y
+coalg-unfold-universal C,γ = lim-terminal {C,γ = C,γ} .snd
+
+-- unique function into final coalg
+coalg-unfold-function : ∀ {ℓ} {S : Container ℓ} → (C,γ : Coalg₀ S) -> (C,γ .fst) -> (M-coalg .fst)
+coalg-unfold-function C,γ y = (coalg-unfold C,γ) .fst y
+
+M-final-coalg : ∀ {ℓ} {S : Container ℓ} -> Final {S = S}
+M-final-coalg {ℓ} {S = S} = M-coalg , λ C,γ → lim-terminal {C,γ = C,γ}
+
+-- final-is-contr : ∀ {ℓ} {S : Container {ℓ}} → isContr (Final {S = S})
```

Comment:
This seems like it should be finished?

Reply:
Yes, I will have a look at that, when I have some time.